security/pkg/pki/util/keycertbundle_test.go

```diff
@@ -343,145 +343,131 @@ func TestNewVerifiedKeyCertBundleFromFile(t *testing.T) {
 
 // Test the root cert expiry timestamp can be extracted correctly.
 func TestExtractRootCertExpiryTimestamp(t *testing.T) {
-	t0 := time.Now()
-	cert, key, err := GenCertKeyFromOptions(CertOptions{
-		Host:         "citadel.testing.istio.io",
-		NotBefore:    t0,
-		TTL:          time.Minute,
-		Org:          "MyOrg",
-		IsCA:         true,
-		IsSelfSigned: true,
-		IsServer:     true,
-		RSAKeySize:   2048,
-	})
-	if err != nil {
-		t.Errorf("failed to gen cert for Citadel self signed cert %v", err)
-	}
-	kb, err := NewVerifiedKeyCertBundleFromPem(cert, key, nil, cert)
-	if err != nil {
-		t.Errorf("failed to create key cert bundle: %v", err)
-	}
 	testCases := []struct {
-		name string
-		ttl  float64
-		time time.Time
+		NotBefore time.Time
+		TTL       time.Duration
 	}{
 		{
-			name: "ttl valid",
-			ttl:  30,
-			time: t0.Add(time.Second * 30),
+			NotBefore: time.Unix(0, 0),
+			TTL:       time.Minute,
 		},
 		{
-			name: "ttl almost expired",
-			ttl:  2,
-			time: t0.Add(time.Second * 58),
+			NotBefore: time.Unix(1721769413, 1000),
+			TTL:       time.Minute * 70,
 		},
 		{
-			name: "ttl just expired",
-			ttl:  0,
-			time: t0.Add(time.Second * 60),
+			NotBefore: time.Unix(2721769413, 1000),
+			TTL:       time.Minute * 10,
 		},
 		{
-			name: "ttl-invalid",
-			ttl:  -30,
-			time: t0.Add(time.Second * 90),
+			NotBefore: time.Unix((1<<31)-2, 1000),
+			TTL:       time.Minute * 10,
 		},
 	}
-	for _, tc := range testCases {
-		t.Run(tc.name, func(t *testing.T) {
+	for i, tc := range testCases {
+		t.Run(fmt.Sprintf("Test case %d", i), func(t *testing.T) {
+			cert, key, err := GenCertKeyFromOptions(CertOptions{
+				Host:         "citadel.testing.istio.io",
+				NotBefore:    tc.NotBefore,
+				TTL:          tc.TTL,
+				Org:          "MyOrg",
+				IsCA:         true,
+				IsSelfSigned: true,
+				IsServer:     true,
+				RSAKeySize:   2048,
+			})
+			if err != nil {
+				t.Errorf("failed to gen cert for Citadel self signed cert %v", err)
+			}
+			// Don't use NewVerifiedKeyCertBundleFromPem because we want to test expired key bundles too
+			kb := NewKeyCertBundleFromPem(cert, key, nil, cert)
+			// will return error if expired
 			expiryTimestamp, _ := kb.ExtractRootCertExpiryTimestamp()
-			// Ignore error; it just indicates cert is expired which we check via `tc.ttl`
-
-			sec := expiryTimestamp - float64(tc.time.Unix())
-			if sec != tc.ttl {
-				t.Fatalf("expected ttl %v, got %v", tc.ttl, sec)
+			expectedExpiryTimestamp := tc.NotBefore.Add(tc.TTL)
+            // One second toleration because x509 cert times have one second of precision
+            tol := time.Second
+			if expiryTimestamp.Sub(expectedExpiryTimestamp).Abs() > tol {
+				t.Errorf("Expected %d and %d to be almost equal", expiryTimestamp.Unix(), expectedExpiryTimestamp.Unix())
```

Comment:
Having more precise TimeBeforeCertExpires means that we have to be more careful with our comparisons since x509 certificates are accurate up to the second.